### DISCUSSION TITLE
ESPHome support?

### DISCUSSION THREAD

**Discussion by netmindz** [netmindz/balboa_GL_ML_spa_control#40] [Ideas]
Would there be any advantage to making an ESPHome custom component rather than the MQTT-based ArduinoHomeAssistant ?

https://esphome.io/components/sensor/custom.html

https://esphome.io/custom/custom_component.html

**Comment by daad919** [comment]
Beste,

we zijn er even tussenuit t.e.m. 15/08/2023.
Na ons verlof zijn we je terug graag van dienst.

Prettige vakantie!

David & Nele

**Reply by netmindz** [reply]
And in English?

**Comment by davewatson91** [comment]
Just my 2c 💯 

I certainly think so. ESPHome makes ESP devices infinitely easier to manage, deploy and update. If it were possible - it would make this much more accessible. I desperately wanted to assist with this project, but I'm so time poor that I haven't had the time to dig in to even get the initial bits going - I even have all the hardware, and just got nowhere :( ESPHome removes a lot of the variables in the setup of the device - if you've physically connected it, then the platform should work. 

😊

**Reply by netmindz** [reply]
Sorry to hear you didn't have any luck. Do you have a working MQTT server setup already for your HomeAssistant install?

**Reply by davewatson91** [reply]
I do, I don't use it much (literally only a couple of legacy things running through it), but it's there. I have maybe ~15 ESPHome devices, and most other things are API based where I can - the MQTT service is for some of my original projects (which I'd do differently now), where I had a rPI running remotely just to switch some relays (which would be better suited for an ESP32 & ESPHome). 

I think my lack of success was a bit around trying to use a MAX485 as well, which I needed to research more to understand what was going on (since everything here was targetted at the automatic RS485 adapters). The doco you've written up now is tons better - I should probably order one of the newer adapters and revisit it 😊

**Reply by netmindz** [reply]
Glad the docs make it clearer. We do also support the older max485 too, just need to connect both the enable pins and define in the code

**Reply by davewatson91** [reply]
I've just got around to finding a few minutes to get this running - I'm not super familiar with coding directly in arduino (am a competent python programmer). Do you have an example of where the enable pins need to be defined for the MAX485? 

I can see the comment in the code for using the spa pin 5 / esp pin 22 for allowing it to work, but not for the actual MAX485 side of it.

**Comment by tmjo** [comment]
I think it would be a good go. Think it would make it really seamless to see when there's a new update, easily update and an just a great way to handle things. If I understood the example I sent earlier right (the Mitsubishi heatpump integration), I think it also gives some benefits similar to custom integrations too. 

But I guess we could wait until stuff is a bit more mature, anyway I'm willing to give it a go and see if we can make it work. 

The whole idea, I think, is that the "library" itself is standalone, and then the esphome integration comes on top of that and uses the library. So there should be no issues for people wanting to use the library as today.

**Reply by netmindz** [reply]
I had a very quick look at how easily I can split the current code into library vs HA and as there is so much state, it's not just a quick refactor, but think would be worth doing. Now the updated AHA can handle only sending updates if the data has changed, we can simplify the code. Think the library will need to populate a struct with all the details and then the ESPHome or AHA code then read the data back out and call the right setters

**Reply by tmjo** [reply]
Sounds like a good idea, a struct with all data is one of the ideas I have for refactoring. I've got something half ready that I tested out some time ago, including removing some Strings and other stuff. I'll see if I can upload it in a separate branch just for the sake of discussion.

**Reply by netmindz** [reply]
Not tested yet, but this at least compiles 😉

https://github.com/netmindz/balboa_GL_ML_spa_control/pull/42

**Reply by netmindz** [reply]
Over to you to try and write the custom ESPHome component? ;)

**Reply by tmjo** [reply]
Great, I'll give it a try. But is it correct to rename it to *.h? h-files should be header files with definitions only, I think the main library should be cpp but then with an h-file which can be included in other use cases. Take a look at [this](https://github.com/SwiCago/HeatPump/tree/master/src) which is the library, a cpp file (logics) and a h-file (definitions) and then the ESPHome component [here ](https://github.com/geoffdavis/esphome-mitsubishiheatpump/blob/master/components/mitsubishi_heatpump/espmhp.h) where it references the Heatpump.h from the library.

Anyways, I'll give it a shot and see. I did some testing trying to rename our sensor.ino to sensor.cpp but it gave me such a headache I hade to stop. Mostly because of lacking skills I think, but still. Mostly related to the other files in our library. Shouldn't be too complicated I believe, but didn't get around to finalize it yet.

I uploaded my [refactoring testing branch](https://github.com/netmindz/balboa_GL_ML_spa_control/tree/send-command-v2-test-refactoring-ideas) just as an exchange of ideas. Not for merging or anything. It compiles, and I've tested most of it, but not commands yet. Could be there are some late night mistakes, and there are still several things I'm not happy with myself. But it could be a way to start a discussion perhaps? I looked at [cribskib's code ](https://github.com/cribskip/esp8266_spa/blob/master/src/esp8266_spa.ino) for inspiration on a lot of it.

**Comment by netmindz** [comment]
I have started putting the basic structure together https://github.com/netmindz/balboa_GL_ML_spa_control/pull/48

**Reply by netmindz** [reply]
Did you see this @tmjo ?

**Reply by tmjo** [reply]
Hi, I noticed, but I didn't find the time to dig around any on it recently. Will for sure look into it though.

**Comment by netmindz** [comment]
I've just pushed code to the ESPHome branch that at lest compiles - doesn't do anything yet, but the scaffolding is complete now @tmjo or anyone else who wants to actually build out the component itself

**Reply by netmindz** [reply]
ok linking fails, think i've pushed a fix for that but change to library being cached I think

**Comment by netmindz** [comment]
Ok, now has a branch that does create a valid ESPHome entry, but none of the actually writing of the balboaGL code to the climate component is done yet.

Not sure with ESPHome if we can expose all the different sensors without needing to define them all one by one the yaml

**Comment by netmindz** [comment]
I now have my new Lilygo T-RSC3 running the ESPHome branch. Doesn't do anything yet so far other than to try and read from the serial port, but that's a start ;)

**Reply by netmindz** [reply]
I now have current temp being returned to ESPHome :)

**Comment by netmindz** [comment]
Progress update:

Docs for writing anything but the most basic components is rather lacking, for example not found any docs for the python script used to link your yaml to the C code.

By looking at a the source of a few other components I'm slowly piecing what we need together.

Ultimately we will want a single component rather than collection of climate, sensor and switch. So will need to do a refactor for that.

I'm also unclear if direct access to HardwareSerial or ESPHomes UART is the most appropriate approach. At the moment, still using HardwareSerial as that's what is used for the regular "sensor" code

If anyone has a tub with 3 ports like mine, you can try and I'm doing and running the more complete original code on one and the new ESPHome on another while we are still working on feature parity

**Comment by netmindz** [comment]
I'm a little concerned that I'm spending huge amounts of time trying to get feature parity between the existing Arduino sensor app and ESPhome component, but interest seems to have gone quiet

You can do OTA with the existing sensor and it already does auto discovery, so really esphome just means your don't need to install vscode+platformio to set your WiFi and MQTT connection details

I'm tempted to park any further development unless I hear back that folk are still interested in this option

**Reply by davewatson91** [reply]
For me, largely it's keeping everything in an easy to manage place (where I also have ~30 other devices). 

I haven't managed to get my setup working in general yet, but also couldn't work out the OTA setup using PlatformIO (this is the first time using PlatformIO). I need to work on getting it going in general, but I feel that ESPHome makes it more accessible to the lamen. 

Also totally respect the time sink, so understand that side too 😄 

I need to spend some time just trying to work out what's going on with mine in general - I have the new T-RSC3, and still can't get any comms with the unit.

**Reply by netmindz** [reply]
Good to hear that there is still interest.

The standard ArduinoOTA just stopped working and not sure why. So it now uses webOTA. So you build and then have to upload via the webpage on port 8080 / update

Bit clunky, but better than nothing

Are you getting data but can't send or nothing at all? The current code in the main branch should at least give read only at the moment. Note: I found A and B were inverted compared to my other rs485 setup

**Reply by netmindz** [reply]
You can also try the https://github.com/netmindz/balboa_GL_ML_spa_control/blob/ESPHome/component.yaml this too, though very much a work in progress

**Reply by netmindz** [reply]
![Screenshot from 2023-11-02 14-43-43](https://github.com/netmindz/balboa_GL_ML_spa_control/assets/442066/a41794ec-8c55-44ea-977c-2043b15c3635)

**Comment by netmindz** [comment]
Hit a snag. Struggling to get ESPHome to respond quickly enough. In my own code I suppress doing a anything other than taking with the tub when trying to send a command, but not so easy to do with ESPHome that has it is own stuff in the main loop for the other components

**Reply by davewatson91** [reply]
This isn't ideal. I wish I had more time on my hands to be able to dig and learn sorry.

**Comment by davewatson91** [comment]
> Good to hear that there is still interest.
> 
> The standard ArduinoOTA just stopped working and not sure why. So it now uses webOTA. So you build and then have to upload via the webpage on port 8080 / update
> 
> Bit clunky, but better than nothing
> 
> Are you getting data but can't send or nothing at all? The current code in the main branch should at least give read only at the moment. Note: I found A and B were inverted compared to my other rs485 setup

I've not got data reading in (A & B were reversed), but not able to send anything still. The only branch that's working for me is the main branch. I haven't played with the ESPHome branch yet - not sure how it needs to be configured. 

Any thoughts on how I can help troubleshoot the inability to send commands?

**Reply by davewatson91** [reply]
I wonder if my issues are related to interpretation of data? I'm only getting a temp, no other valid values... I have a blower, circulation pump & main pump.

**Reply by netmindz** [reply]
I presume you meant to say now rather than not.

The main branch of best,  but with my own T-RSC3 it just fails to send a command to the controller for code that works fine on a regular esp32 + rs485 adapter

**Reply by netmindz** [reply]
Can you share a screenshot of what is showing in home assistant for this device

**Reply by netmindz** [reply]
The ESPHome support at the moment is very much a work in progress, while main is what I'm using day to day for my automations

**Reply by davewatson91** [reply]
Definitely meant to say now 😶‍🌫️ 

![image](https://github.com/netmindz/balboa_GL_ML_spa_control/assets/51277302/6a7e47eb-b654-4e86-a983-a887b310096f)

Hopefully something obvious stands out there to you!

**Reply by netmindz** [reply]
Looks like reading from the tub is working. Target temp should show up in you press your temp+ or temp- on the hot tub. 

If you turn on the pumps, lights etc on the existing panel is the status shown correctly in HA?

The only thing that looks unusual is the FB message that is all ff

For future discussion about using the main branch with this LILYGO T-RSC3, please use the discussion thread for that
